stan/math/prim/mat/prob/bernoulli_logit_glm_lpmf.hpp

```diff
@@ -0,0 +1,150 @@
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+#include <boost/random/bernoulli_distribution.hpp>
+#include <boost/random/variate_generator.hpp>
+#include <cmath>
+#include <string>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Returns the log PMF of the Generalized Linear Model (GLM)
+     * with Bernoulli distribution and logit link function.
+     * If containers are supplied, returns the log sum of the probabilities.
+     * @tparam T_n type of binary vector of dependent variables (labels);
+     * this can also be a single binary value;
+     * @tparam T_x type of the matrix of independent variables (features); this
+     * should be an Eigen::Matrix type whose number of rows should match the 
+     * length of n and whose number of columns should match the length of beta
+     * @tparam T_beta type of the weight vector;
+     * this can also be a single double value;
+     * @tparam T_alpha type of the intercept;
+     * this can either be a vector of doubles of a single double value;
+     * @param n binary vector parameter
+     * @param x design matrix
+     * @param beta weight vector
+     * @param alpha intercept (in log odds)
+     * @return log probability or log sum of probabilities
+     * @throw std::domain_error if theta is infinite.
+     * @throw std::invalid_argument if container sizes mismatch.
+     */
+    template <bool propto, typename T_n, typename T_x, typename T_beta,
+              typename T_alpha>
+    typename return_type<T_x, T_beta, T_alpha>::type
+    bernoulli_logit_glm_lpmf(const T_n &n, const T_x &x, const T_beta &beta,
+                             const T_alpha &alpha) {
+      static const std::string function = "bernoulli_logit_glm_lpmf";
+      typedef typename stan::partials_return_type<T_n, T_x, T_beta,
+                                                  T_alpha>::type
+        T_partials_return;
+
+      using stan::is_constant_struct;
+      using std::exp;
+      using Eigen::Dynamic;
+      using Eigen::Matrix;
+
+      if (!(stan::length(n) && stan::length(x) && stan::length(beta)))
+        return 0.0;
+
+      T_partials_return logp(0.0);
+
+      check_bounded(function, "Vector of dependent variables", n, 0, 1);
+      check_not_nan(function, "Matrix of independent variables", x);
+      check_not_nan(function, "Weight vector", beta);
+      check_not_nan(function, "Intercept", alpha);
+      check_consistent_sizes(function, "Rows in matrix of independent variables",
+                             x.col(0), "Vector of dependent variables",  n);
+      check_consistent_sizes(function, "Columns in matrix of independent variables",
+                             x.row(0), "Weight vector",  beta);
+
+      if (!include_summand<propto, T_x, T_beta, T_alpha>::value)
+        return 0.0;
+
+      const size_t N = x.col(0).size();
+      const size_t M = x.row(0).size();
+      
+      Matrix<double, Dynamic, 1> signs(N,1);
+      {
```

Comment:
You don't really need to scope your local variables like this.  The compiler's really good at dealing with lifespans of variables.

Reply:
That's great! I didn't realise that. :-)